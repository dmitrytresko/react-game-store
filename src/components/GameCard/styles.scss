$largeMediaWidth: 1201px;
$maxMediaWidth: 1024px;
$tabletMediaWidth: 768px;

.game-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 24%;
  min-width: 240px;
  max-width: 290px;
  margin: 0 10px 20px;
  overflow: hidden;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.4s ease;

  @media (min-width: $maxMediaWidth) {
    min-width: 260px;
  }

  @media (min-width: $tabletMediaWidth) {
    min-width: 250px;
  }

  &__cover-wrapper {
    position: relative;
    width: 100%;
    height: 19.5vw;
    min-height: 340px;
    overflow: hidden;
    background: black;
    border-bottom: 2px solid #f8f8f8;
    transition: all 0.4s ease-in-out;

    @media (max-width: $largeMediaWidth) {
      height: 350px;
      min-height: 0;
    }

    @media (max-width: $maxMediaWidth) {
      height: 335px;
      min-height: 0;
    }

    .game-card__cover {
      width: 100%;
      height: 100%;
      opacity: 1;
      transition: opacity 0.4s ease-in-out;
    }

    .game-card__add-btn {
      position: absolute;
      right: 0;
      bottom: 0;
      padding: 5px 20px;
      color: coral;
      font-size: 23px;
      font-family: "Teko", sans-serif;
      background: #f8f8f8;
      border: 2px solid coral;
      border-right: none;
      border-bottom: none;
      border-top-left-radius: 5px;
      outline: none;
      transform: translateY(50px);
      cursor: pointer;
      transition: all 0.4s;

      &:hover {
        color: #f8f8f8;
        background: coral;
        border-color: coral;
      }

      &:active {
        color: coral;
        background: black;
      }

      @media (max-width: $maxMediaWidth) {
        font-size: 20px;
      }
    }

    .game-card__edit-btn {
      position: absolute;
      bottom: 0;
      left: 0;
      padding: 5px 20px;
      color: coral;
      font-size: 23px;
      font-family: "Teko", sans-serif;
      background: #f8f8f8;
      border: 2px solid coral;
      border-bottom: none;
      border-left: none;
      border-top-right-radius: 5px;
      outline: none;
      transform: translateY(50px);
      cursor: pointer;
      transition: all 0.4s;

      &:hover {
        color: #f8f8f8;
        background: coral;
        border-color: coral;
      }

      &:active {
        color: coral;
        background: black;
      }

      @media (max-width: $maxMediaWidth) {
        font-size: 20px;
      }
    }
  }

  &__details-container {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    justify-content: space-between;
    width: 100%;
    min-height: 135px;
    padding: 15px;
    background: black;
    transition: all 0.4s;

    .game-card__info {
      display: flex;
      justify-content: space-between;
      width: 100%;

      .game-card__title {
        max-width: 82%;
        color: coral;
        font-size: 23px;
        font-family: "Teko", sans-serif;

        @media (max-width: $tabletMediaWidth) {
          font-size: 21px;
        }
      }

      .game-card__company {
        color: #f8f8f8;
        font-size: 23px;
        font-family: "Teko", sans-serif;

        @media (max-width: $tabletMediaWidth) {
          font-size: 20px;
        }
      }

      .game-card__rating {
        width: 32px;
        height: 32px;
      }

      .game-card__price {
        color: white;
        font-weight: bold;
        font-size: 24px;
        font-family: "Teko", sans-serif;

        @media (max-width: $tabletMediaWidth) {
          font-size: 21px;
        }
      }
    }

    .game-card__info:last-child {
      align-items: flex-end;
    }

    @media (max-width: $tabletMediaWidth) {
      min-height: 130px;
    }
  }

  &:hover {
    box-shadow: 0 0 10px 3px coral;
    transform: translateY(-5px);
  }

  &:hover &__cover-wrapper {
    border-color: coral;

    .game-card__cover {
      opacity: 0.4;
    }

    .game-card__edit-btn {
      transform: translateY(0);
    }

    .game-card__add-btn {
      transform: translateY(0);
    }
  }

  &:hover &__details-container {
    background-color: #f8f8f8;

    .game-card__title {
      color: black;
    }

    .game-card__company {
      color: dimgray;
    }

    .game-card__price {
      color: coral;
    }
  }
}
