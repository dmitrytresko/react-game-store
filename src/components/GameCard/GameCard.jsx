import { useState, useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { SET_CART_DATA } from "../../redux/actions";
import Modal from "../../elements/Modal/Modal";
import InputText from "../../elements/InputText/InputText";
import "./styles.scss";

const GameCard = ({ gameDetails }) => {
  const dispatch = useDispatch();

  const [modalState, setModalState] = useState({ isOpened: false, editGameClicked: false });

  const isUserLoggedIn = useSelector(state => state.user?.isLogged);
  const isUserAdmin = useSelector(state => state.user?.isAdmin);
  const userCartCount = useSelector(state => state.user?.cartCount);
  const userSelectedItems = useSelector(state => state.user?.selectedItems);

  const addItemToCartHandler = () => {
    if (isUserLoggedIn) {
      dispatch({
        type: SET_CART_DATA,
        payload: {
          newCartCount: userCartCount + 1,
          selectedItems: [
            ...userSelectedItems,
            {
              gameId: gameDetails.id,
              gameName: gameDetails.name,
              gamePrice: gameDetails.price,
              gameCompany: gameDetails.company,
              gameImage: gameDetails.path
            }
          ]
        }
      });
    }
  }

  const editItemHandler = () => {
    console.log(gameDetails);

    setModalState({ isOpened: true, editGameClicked: true });
  }

  const truncateGameName = (name = '', maxLength = 35) => {
    return name.length > maxLength ? `${name.substring(0, maxLength)}â€¦` : name;
  }

  const truncatedName = useMemo(() => truncateGameName(gameDetails.name, 34), [gameDetails.name]);

  return (
    <>
      <div className="game-card">
        <div className="game-card__cover-wrapper">
          <img className="game-card__cover" src={gameDetails.path} alt={gameDetails.name} />
          {isUserLoggedIn &&
            <>
              <button className="game-card__add-btn" onClick={() => addItemToCartHandler(gameDetails)}>Add To Cart</button>
              {isUserAdmin && <button className="game-card__edit-btn" onClick={editItemHandler}>Edit Game</button>}
            </>
          }
        </div>
        <div className="game-card__details-container">
          <div className="game-card__info">
            <h3 className="game-card__title">{truncatedName}</h3>
            <img className="game-card__rating" src={gameDetails.rating} alt="Star rating" />
          </div>
          <div className="game-card__info">
            <p className="game-card__company">{gameDetails.company}</p>
            <p className="game-card__price">${gameDetails.price}</p>
          </div>
        </div>
      </div>

      <Modal opened={modalState.isOpened}
            type={`${modalState.editGameClicked ? "editGame" : "addGame"}`}
            onCloseClick={() => setModalState({ isOpened: false, editGameClicked: false })}>
              <InputText fieldLabel="Name:" fieldName="name" fieldType="text" message="Enter game name here..."></InputText>
              <InputText fieldLabel="Genre:" fieldName="genre" fieldType="text" message="Enter game genre here..."></InputText>
              <InputText fieldLabel="Price:" fieldName="price" fieldType="number" message="Enter game price here..."></InputText>
              <InputText fieldLabel="Company:" fieldName="company" fieldType="text" message="Enter company name here..."></InputText>
              <InputText fieldLabel="Age:" fieldName="age" fieldType="number" message="Enter game age here..."></InputText>
              <InputText fieldLabel="Image:" fieldName="image" fieldType="file" message="Select game image here..."></InputText>
      </Modal>
    </>
  )
}

export default GameCard;
