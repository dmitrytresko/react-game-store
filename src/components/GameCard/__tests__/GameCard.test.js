import React from "react";
import { render } from "@testing-library/react";
import GameCard from "../GameCard";
import { Provider } from 'react-redux';
import fifa21Img from "../assets/img/games/ps/fifa21.png";
import fiveStars from "../assets/img/card-items/five-stars.jpg";

describe("Game card", () => {
  const mockProps = {
    gameDetails: {
      id: 100,
      name: "Game Name",
      company: "Company Name",
      path: fifa21Img,
      rating: fiveStars,
      price: 99.99,
      age: 0,
      genre: "Genre Name",
      metaRating: 99
    },
    openEditGameModalState: jest.fn()
  }

  it("should render Game card component", () => {
    const store = {
      getState: jest.fn(() => ({})),
      dispatch: jest.fn(),
      subscribe: jest.fn(),
    };

    jest.mock('react-redux', () => {
      return {
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn().mockImplementation(() => ({})),
        useDispatch: () => jest.fn(),
      };
    });

    const { container } = render(
      <Provider store={store}>
        <GameCard {...mockProps} />
      </Provider>
    );

    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })
})
