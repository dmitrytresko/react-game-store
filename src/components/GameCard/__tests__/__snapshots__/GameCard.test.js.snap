// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game card should render Game card component 1`] = `
<div>
  <div
    class="game-card"
  >
    <div
      class="game-card__cover-wrapper"
    >
      <img
        alt="Game Name"
        class="game-card__cover"
        src="test-file-stub"
      />
    </div>
    <div
      class="game-card__details-container"
    >
      <div
        class="game-card__info"
      >
        <h3
          class="game-card__title"
        >
          Game Name
        </h3>
        <img
          alt="Star rating"
          class="game-card__rating"
          src="test-file-stub"
        />
      </div>
      <div
        class="game-card__info"
      >
        <p
          class="game-card__company"
        >
          Company Name
        </p>
        <p
          class="game-card__price"
        >
          $
          99.99
        </p>
      </div>
    </div>
  </div>
</div>
`;
